File: android/app/src/main/java/com/dontworry/app/domain/search/Bm25Scorer.kt
package com.dontworry.app.domain.search

import kotlin.math.ln

object Bm25Scorer {
    private const val K1 = 1.2
    private const val B = 0.75

    fun score(
        queryTerms: List<String>,
        termFrequencies: Map<String, Int>,
        docLength: Int,
        averageDocLength: Double,
        docFrequencyProvider: (String) -> Int,
        totalDocuments: Int
    ): Double {
        if (queryTerms.isEmpty() || docLength <= 0 || totalDocuments <= 0) return 0.0
        val safeAvg = if (averageDocLength <= 0.0) 1.0 else averageDocLength

        var score = 0.0
        for (term in queryTerms) {
            val tf = termFrequencies[term] ?: 0
            if (tf <= 0) continue

            val df = docFrequencyProvider(term)
            val idf = ln(((totalDocuments - df + 0.5) / (df + 0.5)) + 1.0)
            val numerator = tf * (K1 + 1.0)
            val denominator = tf + K1 * (1.0 - B + B * (docLength / safeAvg))
            score += idf * (numerator / denominator)
        }

        return score
    }
}

File: ios/DontWorryiOS/Domain/Bm25Scorer.swift
import Foundation

enum Bm25Scorer {
    private static let k1 = 1.2
    private static let b = 0.75

    static func score(
        queryTerms: [String],
        termFrequencies: [String: Int],
        docLength: Int,
        averageDocLength: Double,
        docFrequencyProvider: (String) -> Int,
        totalDocuments: Int
    ) -> Double {
        guard !queryTerms.isEmpty, docLength > 0, totalDocuments > 0 else {
            return 0.0
        }

        let safeAvg = averageDocLength <= 0 ? 1.0 : averageDocLength
        var score = 0.0

        for term in queryTerms {
            let tf = termFrequencies[term] ?? 0
            if tf <= 0 { continue }

            let df = docFrequencyProvider(term)
            let idf = log(((Double(totalDocuments - df) + 0.5) / (Double(df) + 0.5)) + 1.0)
            let numerator = Double(tf) * (k1 + 1.0)
            let denominator = Double(tf) + k1 * (1.0 - b + b * (Double(docLength) / safeAvg))
            score += idf * (numerator / denominator)
        }

        return score
    }
}
